script/modelling/maxcut.jl

```diff
@@ -0,0 +1,169 @@
+
+@objective(linear_max_cut, Max, dot(w, z))
+
+optimize!(linear_max_cut)
+
+@test objective_value(linear_max_cut) ≈ 6.0
+@show objective_value(linear_max_cut);
+
+x_linear = value.(x)
+@show x_linear;
+@show [(i,j) for i in 1:n-1 for j in i+1:n if JuMP.value.(z)[i,j] > 0.5];
+
+#' Visualizing a solution
+
+nodecolor = [colorant"lightseagreen", colorant"orange"]
+all_node_colors = [nodecolor[round(Int, xi + 1)]  for xi in x_linear]
+GraphPlot.gplot(g, nodelabel=1:6, nodefillc=all_node_colors)
+
+# membership color
+nodefillc = nodecolor[membership]
+gplot(g, nodefillc=nodefillc)
+
+#' ### Semi-definite formulation
+
+#' The maximum cut problem can also be formulated as a quadratic optimization model.
+#' If the solution vector is composed of $\{-1, 1\}$ for belonging to the subset or not.
+#' Then, the product $x_i x_j$ is $-1$ if i and j are in different subsets, and 1 otherwise.
+#' This implies $1 - x_i x_j$ is $0$ if i and j are in the same subset, and 2 otherwise.
+#' Since each pair is counted twice with (i,j) and (j,i), we need to divide it by 4:
+#' $\frac{w_{ij}}{4} (1 - x_i x_j)$ is $0$ if i and j are in the same subset, and $z_{ij}$ otherwise.
+
+#' The resulting model is given by:
+#' $$
+#' \begin{align*}
+#' \max_{x} & \frac{w_{ij}}{4} (1 - x_i x_j) \\
+#' \text{s.t. } & x_i \in \{-1, 1\} \forall i \in V
+#' \end{align*}
+#' $$
+
+#' Introducing a matrix variable $Y_{ij} = x_i x_j$,
+#' we can equivalently reformulate the above quatratic problem as follows:
+#' $$
+#' \begin{align*}
+#' \max_{x,Y} & \frac{w_{ij}}{4} (1 - Y_{ij}) \\
+#' \text{s.t. } & Y_{ii} = 1 \forall i \in V \\
+#'              & Y = x x^T
+#' \end{align*}
+#' $$
+
+#' The constraint $Y = x x^T$ requires $Y$ to be a Positive Semi-Definite matrix
+#' of rank one. This optimization problem is fully equivalent to the initial
+#' quadratic problem and hard to solve.
+#' A possible relaxation is to remove the rank-1, yielding the following problem:
+#' $$
+#' \begin{align*}
+#' \max_{Y} & \frac{w_{ij}}{4} (1 - Y_{ij}) \\
+#' \text{s.t. } & Y_{ii} = 1 \forall i \in V \\
+#'              & Y \in \mathcal{S}_n^+
+#' \end{align*}
+#' $$
+
+#' with $\mathcal{S}_n^+$ the set of positive semi-definite matrices.
+#' Such model can be implemented in JuMP and solved using an underlying solver
+#' which supports semi-definite optimization.
+
+sdp_max_cut = Model(optimizer_with_attributes(SCS.Optimizer, "verbose" => 0))
+
+@variable(sdp_max_cut, Y[1:n,1:n] in PSDCone())
+@constraint(sdp_max_cut, [i = 1:n], Y[i,i] == 1)
+@objective(sdp_max_cut, Max, 1/4 * sum(w[i,j] * (1 - Y[i,j]) for i in 1:n for j in 1:n))
+
+optimize!(sdp_max_cut)
+
+@test objective_value(linear_max_cut) >= 6.0
+@show objective_value(linear_max_cut);
+
+#' Re-computing the vector:
+#' $Y$ approximates $x x^T$, we can compute the estimate by randomized rounding.
+#' TODO add paper.
+
+F = svd(value.(Y))
+xhat_unnormalized = F.U[:,1]
```

Comment:
This is like using (1, 0, 0,..., 0) as random vector. I would rather use an actual random one

Reply:
got it, following the paper -> the random vector has to be divided by its L2-norm right?

Reply:
ok no need to normalize, just taking the sign of the dot product

Reply:
I tried multiplying with an actual random and get wrong results (not sure about the bound though). Could the bad quality be linked with the fact that there is one negative random value, by 3e-8, this feels a bit large already. How does one tighten the tolerance on SCS?

Reply:
namale, with F the svd of Y:
```
julia> F.U - F.V
6×6 Array{Float64,2}:
  4.16334e-17   2.22045e-16  -9.23093e-9   0.509028    -1.53147e-9  -4.09383e-9
  1.11022e-16  -1.11022e-16  -2.28702e-9   0.00410811  -3.49982e-9  -9.71985e-9
  0.0           8.32667e-17   4.91655e-9  -0.361691     4.1682e-9   -8.63571e-9
  0.0           1.38778e-16  -6.81818e-9   0.368006    -2.56051e-9  -2.6518e-9
 -1.66533e-16  -3.33067e-16  -2.7614e-8    1.66857      4.81733e-9  -6.19453e-9
  5.55112e-17   1.94289e-16   1.20126e-8  -0.830955    -1.45271e-8   2.67845e-9
```

Some elements are huge

Reply:
thanks. Do you have a resource we could point to in the tutorial for the first part? Most lectures and papers seem to point at Cholesky directly, since you're supposed to have a PSD Y